File: backend/app/schemas/closing.py
import uuid
from typing import Optional
from datetime import datetime
from pydantic import BaseModel, ConfigDict
from app.models.monthly_closing import ClosingStatus


class ClosingBase(BaseModel):
    property_id: uuid.UUID
    year_month: str


class ClosingResponse(ClosingBase):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    user_id: uuid.UUID
    total_revenue: float
    total_expenses: float
    net_result: float
    total_nights: int
    total_bookings: int
    depreciation_value: float
    cleaning_total: float
    platform_fee_total: float
    other_expenses: float
    status: ClosingStatus
    closed_at: Optional[datetime] = None
    notes: Optional[str] = None
    created_at: datetime
    updated_at: Optional[datetime] = None


class ClosingGenerateRequest(BaseModel):
    property_id: uuid.UUID
    year_month: str


class ClosingNotesUpdate(BaseModel):
    notes: str

File: frontend/src/api/closing.ts
import apiClient from './client'

export interface ClosingResponse {
  id: string
  property_id: string
  year_month: string
  total_revenue: number
  total_expenses: number
  net_result: number
  total_nights: number
  total_bookings: number
  depreciation_value: number
  cleaning_total: number
  platform_fee_total: number
  other_expenses: number
  status: 'DRAFT' | 'CLOSED' | 'CANCELLED'
  notes?: string
  closed_at?: string
  created_at: string
  updated_at?: string
}

export interface ClosingListResponse {
  data: ClosingResponse[]
  meta?: {
    total: number
    page: number
    per_page: number
    total_pages: number
  }
}

export const closingApi = {
  list: async (property_id?: string): Promise<ClosingListResponse> => {
    const response = await apiClient.get<ClosingListResponse>('/closing', {
      params: property_id ? { property_id } : undefined,
    })
    return response.data
  },

  generate: async (data: { property_id: string; year_month: string }): Promise<ClosingResponse> => {
    const response = await apiClient.post<{ data: ClosingResponse }>('/closing/generate', data)
    return response.data.data
  },

  get: async (property_id: string, year_month: string): Promise<ClosingResponse> => {
    const response = await apiClient.get<{ data: ClosingResponse }>(`/closing/${property_id}/${year_month}`)
    return response.data.data
  },

  close: async (property_id: string, year_month: string): Promise<ClosingResponse> => {
    const response = await apiClient.post<{ data: ClosingResponse }>(`/closing/${property_id}/${year_month}/close`)
    return response.data.data
  },

  reopen: async (property_id: string, year_month: string): Promise<ClosingResponse> => {
    const response = await apiClient.post<{ data: ClosingResponse }>(`/closing/${property_id}/${year_month}/reopen`)
    return response.data.data
  },

  exportCsv: async (property_id: string, year_month: string): Promise<void> => {
    const response = await apiClient.get(`/closing/${property_id}/${year_month}/export`, {
      responseType: 'blob',
    })
    const url = window.URL.createObjectURL(new Blob([response.data]))
    const link = document.createElement('a')
    link.href = url
    link.setAttribute('download', `fechamento-${property_id}-${year_month}.csv`)
    document.body.appendChild(link)
    link.click()
    link.remove()
  },
}
